### ISSUE TITLE
line breaks in dialogue

### ISSUE THREAD

**Issue by incandescentman** [rnkn/fountain-mode#40]
According to the fountain syntax, you can include empty lines within Dialogue by doing `<RET><space<space><RET>`. From [the documentation](http://fountain.io/syntax#section-br):

![line breaks in dialogue](http://i.imgur.com/cP6a6df.png)

...However, fountain-mode does not seem to be exporting this correctly. The following text in fountain-mode:

    VOICE
    From the secret palaces of the Orient....from the modern masters at Hollywood Vividcolor ...comes the most alluring, most entrapping, most reality-altering lipstick ever.  
    <space><space>
    Hypnotic Honeydew. It's not just a lipstick...it's pharmacological entrapment. Try it on that one special guy... He'll never know what hit him.
    <space><space>  
    One kiss and he's helpless; one kiss and he's yours forever!
        <space><space>
     Remember, if you don't use it on, someone else will! Act now while there's still time!
         <space><space>
    Hypnotic Honeydew. Try it. Your life will never be the same again.

...exports HTML that looks like this:
![HTML](http://i.imgur.com/GR3i1bI.png)

...and exports PDF that looks like this:

![PDF](http://i.imgur.com/2NTWNYO.png)

Is there a way to make this work?

**Comment by rnkn** [comment]
The exporter I wrote (Fountain → HTML → PDF via Prince) is kinda junk. HTML is not a terribly good intermediary format for something intended to be printed on paper, and the development direction of Prince doesn't look like it will ever make it "good enough." I'm planning to remove it in version 2.

The best way to export I've found is [afterwriting CLI](https://github.com/ifrost/afterwriting-labs/blob/master/docs/clients.md) for node.js instead. It does a much better job. You can install that and set `fountain-export-shell-script` to something like the following:
~~~
afterwriting --config ~/.config/afterwriting/config.json --source %s --pdf --overwrite
~~~

Using this method requires setting export options in afterwriting's `config.json` instead of using Fountain mode's custom options. (Maybe one day afterwriting CLI will feature command line flags for these options, but we'll have to wait and see.)

**Comment by incandescentman** [comment]
Awesome, thanks. And how do I call the new `fountain-export-shell-script` from within fountain-mode?

**Comment by rnkn** [comment]
Ah! Terribly sorry, I've given you instructions based on things I've put into v2.0 but not into master branch yet.

If you paste the below code into `fountain-mode.el`, overwriting the functions of the same names, then <kbd>C-M-x</kbd> on each (or kill Emacs), you should be up and running. Then just <kbd>C-c C-e C-e</kbd> will run the afterwriting script.

~~~emacs
(defcustom fountain-export-default-command
  'fountain-export-shell-script
  "\\<fountain-mode-map>Default function to call with \\[fountain-export-default]."
  :type '(radio (function-item fountain-export-shell-script)
                (function-item fountain-export-buffer-to-html))
  :group 'fountain-export)

(defcustom fountain-export-shell-script
  "afterwriting --config ~/.config/afterwriting/config.json --source %s --pdf --overwrite"
  "Shell command string to convert Fountain source to ouput.
\"%s\" will be substituted with `buffer-file-name'"
  :type 'string
  :group 'fountain-export)

(defun fountain-export-shell-script (&optional buffer)
  "Call shell script defined in `fountain-export-shell-script'."
  (interactive)
  (let* ((buffer (or buffer (current-buffer)))
         (file (shell-quote-argument (buffer-file-name buffer)))
         (command (format fountain-export-shell-script file)))
    (async-shell-command command "*Fountain Export Process*")))
~~~

I really thought this was part of the released version, which is a good sign to push a new release. I'll do that within the next couple of days. Sorry about the wild goose chase.

**Comment by incandescentman** [comment]
Awesome! OK I eval'd that code, and doing <kbd>C-c C-e C-e</kbd> or `fountain-export-default` still seems to be going through HTML:

    prince: Loading document...
    prince: loading HTML input: /Users/im/test.html
    prince: /Users/jay/Downloads/hypnotic-honeydew.html:249: error: Tag section invalid
    prince: Converting document...
    prince: used font: Courier, Regular
    prince: Finished: success

But no rush at all, I can just wait for the new release. Just wanted to update you with my results.

**Comment by incandescentman** [comment]
By the way, it would still be great to make the line breaks in dialogue work correctly just for HTML export. If you look at the HTML above, you'll see that fountain export to HTML concatenates all the dialogue together in the exported HTML, instead of leaving empty lines as per the fountain spec.

**Comment by rnkn** [comment]
I've pushed a half-fix for this here a7ba3d7.

It's only halfway there because HTML will always collapse whitespace, so you'll only get a single line break instead of a blank line, unless there's a non-breaking space (`&nbsp;`) character in between.

Luckily fountain-mode allows inserting a non-breaking space with `\ ` (backslash space), so you can do the following:
~~~
VOICE
From the secret palaces of the Orient....from the modern masters at Hollywood Vividcolor ...comes the most alluring, most entrapping, most reality-altering lipstick ever.  
\<space>
Hypnotic Honeydew. It's not just a lipstick...it's pharmacological entrapment. Try it on that one special guy... He'll never know what hit him.
~~~
Caveat: this is not in the official spec and not supported by afterwriting.

I don't really want to write the code to look for the double-space and convert to non-breaking because the current exporting code is 100% replaced in the `develop` branch, which will be merged into master soon.

**Comment by incandescentman** [comment]
Excellent. The above solution `\ ` works perfectly for exporting to HTML!

Once the develop branch is merged into master, will there be a way to export both to HTML (as per above) and to PDF (using `afterwriting`) from the same `.fountain` source file that will (1) leave line breaks in dialogue intact in export to HTML, but (2) not leave backslashes in PDF export through `afterwriting`?

**Comment by rnkn** [comment]
Yes to both. Although the develop branch as is does not include any HTML export function, so if you clone & merge it will just be disappointing.

I'd planned to release v2.0 with just export via shell command but it looks like that would screw you around, so it may be a week or so until I get the new HTML export code done and included.

**Comment by incandescentman** [comment]
Awesome!

**Comment by rnkn** [comment]
This is fixed in the MELPA version but not yet tagged for release on MELPA-stable.

**Comment by rnkn** [comment]
v2.0 is tagged and will be on MELPA-stable soon.
